### PR TITLE
fixed recovery hang and undefined set_close method in aiokafka

### PR DESCRIPTION
**Pull request by patkivikram**
*Note*: Before submitting this pull request, please review our [contributing
guidelines](https://faust.readthedocs.io/en/master/contributing.html).

## Description

Probable fix for https://github.com/faust-streaming/faust/issues/21

### CODE REVIEW

File: faust/transport/drivers/aiokafka.py [hunk 1]
```diff
@@ -487,8 +487,13 @@ def on_generation_id_known(self) -> None:
     def close(self) -> None:
         """Close consumer for graceful shutdown."""
         if self._consumer is not None:
-            self._consumer.set_close()
-            self._consumer._coordinator.set_close()
+            self._consumer._closed = True
```
Comment (sdrichter) [inline faust/transport/drivers/aiokafka.py:490]: I don't think accessing a variable marked private is a good idea

File: faust/transport/drivers/aiokafka.py [hunk 2]
```diff
@@ -487,8 +487,13 @@ def on_generation_id_known(self) -> None:
...
+            asyncio.run_coroutine_threadsafe(
+                self._consumer._client.close(), self.app.loop
```
Comment (sdrichter) [inline faust/transport/drivers/aiokafka.py:492]: Can we avoid accessing the private variable _client?

File: faust/transport/drivers/aiokafka.py [hunk 3]
```diff
@@ -487,8 +487,13 @@ def on_generation_id_known(self) -> None:
...
+            )
+            asyncio.run_coroutine_threadsafe(
+                self._consumer._coordinator.close(), self.app.loop
```
Comment (sdrichter) [inline faust/transport/drivers/aiokafka.py:495]: Can we avoid accessing the private variable _coordinator?

File: tests/unit/transport/drivers/test_aiokafka.py [hunk 4]
```diff
@@ -918,8 +918,8 @@ def test_start_coordinator_span(self, *, cthread):
     def test_close(self, *, cthread, _consumer):
         cthread._consumer = _consumer
         cthread.close()
-        _consumer.set_close.assert_called_once_with()
-        _consumer._coordinator.set_close.assert_called_once_with()
+        _consumer._closed is True
+        _consumer._coordinator.close.assert_called_once_with()
```
Comment (sdrichter) [inline tests/unit/transport/drivers/test_aiokafka.py:922]: Same with the private variables

Comment (marcosschroh) [reply tests/unit/transport/drivers/test_aiokafka.py:922]: I think should be `assert _consumer._closed`

File: faust/tables/recovery.py [hunk 5]
```diff
@@ -545,22 +549,17 @@ def _estimated_active_remaining_secs(self, remaining: float) -> Optional[float]:
         else:
             return None
 
-    async def _wait(self, coro: WaitArgT) -> None:
-        wait_result = await self.wait_first(
-            coro,
-            self.signal_recovery_reset,
-            self.signal_recovery_start,
-        )
+    async def _wait(self, coro: WaitArgT, timeout=None) -> None:
```
Comment (marcosschroh) [inline faust/tables/recovery.py:552]: let's use the annotation `Seconds` for timeout (`from mode.utils.times import` )

Comment (patkivikram) [reply faust/tables/recovery.py:552]: using int

File: faust/tables/recovery.py [hunk 6]
```diff
@@ -371,47 +363,60 @@ async def _restart_recovery(self) -> None:
...
-                    if active_offsets[tp] > active_highwaters[tp]:
-                        raise ConsistencyError(
-                            E_PERSISTED_OFFSET.format(
-                                tp,
-                                active_offsets[tp],
-                                active_highwaters[tp],
-                            ),
-                        )
+                    if active_offsets[tp] and active_highwaters[tp]:
+                        if active_offsets[tp] > active_highwaters[tp]:
```
[the remark this answers is not in the document]
Comment (patkivikram) [reply faust/tables/recovery.py:389]: fixed

File: faust/transport/drivers/aiokafka.py (hunk 2 above)
Comment (patkivikram) [reply faust/transport/drivers/aiokafka.py:492]: We will have to create those methods in aiokafka. robinhood-aiokafka had public methods to close

File: faust/transport/drivers/aiokafka.py (hunk 3 above)
Comment (patkivikram) [reply faust/transport/drivers/aiokafka.py:495]: we will add those in aiokafka. They exist in robinhood's version

**Review by marcosschroh — APPROVED** [review]

**Pull request merged**